pkg/sql/schemachanger/schemachanger_test: TestConcurrentDeclarativeSchemaChanges failed

pkg/sql/schemachanger/schemachanger_test.TestConcurrentDeclarativeSchemaChanges [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10890947?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10890947?buildTab=artifacts#/) on master @ [b19302fd436f7e8667b289677723e337d333bb50](https://github.com/cockroachdb/cockroach/commits/b19302fd436f7e8667b289677723e337d333bb50):


```
=== RUN   TestConcurrentDeclarativeSchemaChanges
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/3e4c90800e6eae84b5bf83507e7ef6e8/logTestConcurrentDeclarativeSchemaChanges1906842654
    test_log_scope.go:81: use -show-logs to present logs inline
    schemachanger_test.go:169: 
        	Error Trace:	pkg/sql/schemachanger/schemachanger_test_test/pkg/sql/schemachanger/schemachanger_test.go:169
        	Error:      	Should not be zero, but was 0
        	Test:       	TestConcurrentDeclarativeSchemaChanges
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/3e4c90800e6eae84b5bf83507e7ef6e8/logTestConcurrentDeclarativeSchemaChanges1906842654
--- FAIL: TestConcurrentDeclarativeSchemaChanges (3.74s)
```
The failing assertion is
```
	// The ALTER PRIMARY KEY schema change must have been blocked.
	require.NotZero(t, alterPrimaryKeyBlockedCounter.Load())
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConcurrentDeclarativeSchemaChanges.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29694

pkg/sql/schemachanger/schemachanger_test.TestConcurrentDeclarativeSchemaChanges [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10976727?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10976727?buildTab=artifacts#/) on master @ [30acaf93015008a672667c0d3fdc00ac95f446a6](https://github.com/cockroachdb/cockroach/commits/30acaf93015008a672667c0d3fdc00ac95f446a6):


```
=== RUN   TestConcurrentDeclarativeSchemaChanges
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/3e4c90800e6eae84b5bf83507e7ef6e8/logTestConcurrentDeclarativeSchemaChanges145910907
    test_log_scope.go:81: use -show-logs to present logs inline
    schemachanger_test.go:169: 
        	Error Trace:	pkg/sql/schemachanger/schemachanger_test_test/pkg/sql/schemachanger/schemachanger_test.go:169
        	Error:      	Should not be zero, but was 0
        	Test:       	TestConcurrentDeclarativeSchemaChanges
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/3e4c90800e6eae84b5bf83507e7ef6e8/logTestConcurrentDeclarativeSchemaChanges145910907
--- FAIL: TestConcurrentDeclarativeSchemaChanges (4.77s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #106842 pkg/sql/schemachanger/schemachanger_test: TestConcurrentDeclarativeSchemaChanges failed [C-test-failure O-robot T-sql-foundations branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConcurrentDeclarativeSchemaChanges.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/schema_change_plan_node.go
+++ b/pkg/sql/schema_change_plan_node.go
@@ -184,8 +184,8 @@ func (p *planner) waitForDescriptorSchemaChanges(
 	ctx context.Context, descID descpb.ID, scs SchemaChangerState,
 ) error {
 
-	if knobs := p.ExecCfg().DeclarativeSchemaChangerTestingKnobs; knobs != nil &&
-		knobs.BeforeWaitingForConcurrentSchemaChanges != nil {
+	knobs := p.ExecCfg().DeclarativeSchemaChangerTestingKnobs
+	if knobs != nil && knobs.BeforeWaitingForConcurrentSchemaChanges != nil {
 		knobs.BeforeWaitingForConcurrentSchemaChanges(scs.stmts)
 	}
 
@@ -199,7 +199,6 @@ func (p *planner) waitForDescriptorSchemaChanges(
 	// Wait for the descriptor to no longer be claimed by a schema change.
 	start := timeutil.Now()
 	logEvery := log.Every(10 * time.Second)
-	var wasBlocked bool
 	for r := retry.StartWithCtx(ctx, base.DefaultRetryOptions()); r.Next(); {
 		now := p.ExecCfg().Clock.Now()
 		var isBlocked bool
@@ -218,9 +217,7 @@ func (p *planner) waitForDescriptorSchemaChanges(
 		}); err != nil {
 			return err
 		}
-		if isBlocked {
-			wasBlocked = true
-		} else {
+		if !isBlocked {
 			break
 		}
 		if logEvery.ShouldLog() {
@@ -229,11 +226,9 @@ func (p *planner) waitForDescriptorSchemaChanges(
 					" waited %v so far", descID, timeutil.Since(start),
 			)
 		}
-	}
-
-	if knobs := p.ExecCfg().DeclarativeSchemaChangerTestingKnobs; knobs != nil &&
-		knobs.AfterWaitingForConcurrentSchemaChanges != nil {
-		knobs.AfterWaitingForConcurrentSchemaChanges(scs.stmts, wasBlocked)
+		if knobs != nil && knobs.WhileWaitingForConcurrentSchemaChanges != nil {
+			knobs.WhileWaitingForConcurrentSchemaChanges(scs.stmts)
+		}
 	}
 
 	log.Infof(

--- a/pkg/sql/schemachanger/BUILD.bazel
+++ b/pkg/sql/schemachanger/BUILD.bazel
@@ -57,7 +57,6 @@ go_test(
         "//pkg/sql/sqltestutils",
         "//pkg/testutils",
         "//pkg/testutils/serverutils",
-        "//pkg/testutils/skip",
         "//pkg/testutils/sqlutils",
         "//pkg/util",
         "//pkg/util/ctxgroup",

--- a/pkg/sql/schemachanger/scexec/testing_knobs.go
+++ b/pkg/sql/schemachanger/scexec/testing_knobs.go
@@ -27,9 +27,9 @@ type TestingKnobs struct {
 	// for concurrent schema changes to finish.
 	BeforeWaitingForConcurrentSchemaChanges func(stmts []string)
 
-	// AfterWaitingForConcurrentSchemaChanges is called at the end of waiting
+	// WhileWaitingForConcurrentSchemaChanges is called while waiting
 	// for concurrent schema changes to finish.
-	AfterWaitingForConcurrentSchemaChanges func(stmts []string, wasBlocked bool)
+	WhileWaitingForConcurrentSchemaChanges func(stmts []string)
 
 	// OnPostCommitPlanError is called whenever the schema changer job returns an
 	// error on building the state or on planning the stages.

--- a/pkg/sql/schemachanger/schemachanger_test.go
+++ b/pkg/sql/schemachanger/schemachanger_test.go
@@ -38,7 +38,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sqltestutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
@@ -58,7 +57,6 @@ import (
 // schema changes operating on the same descriptors are performed serially.
 func TestConcurrentDeclarativeSchemaChanges(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 106732, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	ctx, cancel := context.WithCancel(context.Background())
@@ -99,9 +97,9 @@ func TestConcurrentDeclarativeSchemaChanges(t *testing.T) {
 	var params base.TestServerArgs
 	params.Knobs = base.TestingKnobs{
 		SQLDeclarativeSchemaChanger: &scexec.TestingKnobs{
-			AfterWaitingForConcurrentSchemaChanges: func(stmts []string, wasBlocked bool) {
+			WhileWaitingForConcurrentSchemaChanges: func(stmts []string) {
 				for _, stmt := range stmts {
-					if wasBlocked && strings.Contains(stmt, "ALTER PRIMARY KEY") {
+					if strings.Contains(stmt, "ALTER PRIMARY KEY") {
 						alterPrimaryKeyBlockedCounter.Add(1)
 						return
 					}
@@ -160,6 +158,14 @@ func TestConcurrentDeclarativeSchemaChanges(t *testing.T) {
 		}
 		wg.Done()
 	}()
+
+	// The ALTER PRIMARY KEY schema change must block.
+	testutils.SucceedsSoon(t, func() error {
+		if alterPrimaryKeyBlockedCounter.Load() == 0 {
+			return errors.New("waiting for concurrent schema change to block")
+		}
+		return nil
+	})
 
 	// Unblock the create index job.
 	continueNotif <- struct{}{}

